ios-base/Networking/Services/AuthenticationServices.swift

```diff
@@ -9,168 +9,190 @@
-    ) { [weak self] (result: Result<User?, Error>, responseHeaders: [AnyHashable: Any]) in
-      switch result {
-      case .success(let user):
-        if self?.saveUserSession(user, headers: responseHeaders) ?? false {
-          completion(.success(()))
-        } else {
-          completion(.failure(AuthError.userSessionInvalid))
-        }
-      case .failure(let error):
-        completion(.failure(error))
+    )
+    switch response.result {
+    case .success(let user):
+      if
+        let user,
+        await self.saveUserSession(user.data, headers: response.responseHeaders)
+      {
+        return .success(user)
+      } else {
+        return .failure(AuthError.mapping)
       }
+    case .failure:
+      return .failure(AuthError.login)
     }
   }
   
   /// Example Upload via Multipart requests.
   /// TODO: rails base backend not supporting multipart uploads yet
-  func signup(
+  @discardableResult func signup(
     email: String,
     password: String,
-    avatar: UIImage,
-    completion: @escaping (Result<User, Error>) -> Void
-  ) {
+    avatar: UIImage
+  ) async -> Result<UserData, AuthError> {
+    
     guard let picData = avatar.jpegData(compressionQuality: 0.75) else {
       let pictureDataError = App.error(
         domain: .generic,
         localizedDescription: "Multipart image data could not be constructed"
       )
-      completion(.failure(pictureDataError))
-      return
+      return .failure(AuthError.mapping)
     }
     
     // Mixed base64 encoded and multipart images are supported
     // in the [MultipartMedia] array
     // Example: `let image2 = Base64Media(key: "user[image]", data: picData)`
     // Then: media [image, image2]
-    let image = MultipartMedia(key: "user[avatar]", data: picData)
-
+    let image = MultipartMedia(fileName: "\(email)_image", key: "user[avatar]", data: picData)
+    
     let endpoint = AuthEndpoint.signUp(
       email: email,
       password: password,
       passwordConfirmation: password,
       picture: nil
     )
-
-    apiClient.multipartRequest(
-      endpoint: endpoint, paramsRootKey: "", media: [image]
-    ) { [weak self] (result: Result<User?, Error>, responseHeaders: [AnyHashable: Any]) in
-      switch result {
-      case .success(let user):
-        if
-          let user = user,
-          self?.saveUserSession(user, headers: responseHeaders) ?? false
-        {
-          completion(.success(user))
-        } else {
-          completion(.failure(AuthError.userSessionInvalid))
-        }
-      case .failure(let error):
-        completion(.failure(error))
+    
+    let response: RequestResponse<UserData> = await apiClient.multipartRequest(
+      endpoint: endpoint,
+      paramsRootKey: "",
+      media: [image])
+    
+    switch response.result {
+    case .success(let user):
+      if
+        let user,
+        await self.saveUserSession(user.data, headers: response.responseHeaders)
```

Comment:
why do we need to await here if `saveUserSession` is not `async`?

Reply:
This is removed on the second PR